apps/emqx_bridge_snowflake/src/emqx_bridge_snowflake_connector.erl

```diff
@@ -0,0 +1,959 @@
+    destroy_action(ActionResId, ActionState),
+    ConnState = ConnState0#{installed_actions := InstalledActions},
+    {ok, ConnState};
+on_remove_channel(_ConnResId, ConnState, _ActionResId) ->
+    {ok, ConnState}.
+
+-spec on_get_channels(connector_resource_id()) ->
+    [{action_resource_id(), action_config()}].
+on_get_channels(ConnResId) ->
+    emqx_bridge_v2:get_channels_for_connector(ConnResId).
+
+-spec on_get_channel_status(
+    connector_resource_id(),
+    action_resource_id(),
+    connector_state()
+) ->
+    ?status_connected | ?status_disconnected.
+on_get_channel_status(
+    _ConnResId,
+    ActionResId,
+    _ConnState = #{installed_actions := InstalledActions}
+) when is_map_key(ActionResId, InstalledActions) ->
+    ActionState = maps:get(ActionResId, InstalledActions),
+    action_status(ActionResId, ActionState);
+on_get_channel_status(_ConnResId, _ActionResId, _ConnState) ->
+    ?status_disconnected.
+
+-spec on_query(connector_resource_id(), query(), connector_state()) ->
+    {ok, _Result} | {error, _Reason}.
+on_query(
+    _ConnResId, {ActionResId, Data}, #{installed_actions := InstalledActions} = _ConnState
+) when
+    is_map_key(ActionResId, InstalledActions)
+->
+    case InstalledActions of
+        %% #{ActionResId := #{mode := direct} = _ActionState} ->
+        %%     {ok, todo};
+        #{ActionResId := #{mode := aggregated} = ActionState} ->
+            run_aggregated_action([Data], ActionState)
+    end;
+on_query(
+    _ConnResId,
+    #insert_report{action_res_id = ActionResId, opts = Opts},
+    #{installed_actions := InstalledActions} = _ConnState
+) when
+    is_map_key(ActionResId, InstalledActions)
+->
+    #{mode := aggregated, http := HTTPClientConfig} = maps:get(ActionResId, InstalledActions),
+    insert_report_request(ActionResId, Opts, HTTPClientConfig);
+on_query(_ConnResId, Query, _ConnState) ->
+    {error, {unrecoverable_error, {invalid_query, Query}}}.
+
+-spec on_batch_query(connector_resource_id(), [query()], connector_state()) ->
+    {ok, _Result} | {error, _Reason}.
+on_batch_query(_ConnResId, [{ActionResId, _} | _] = Batch0, #{installed_actions := InstalledActions}) when
+    is_map_key(ActionResId, InstalledActions)
+->
+    case InstalledActions of
+        %% #{ActionResId := #{mode := direct} = _ActionState} ->
+        %%     {ok, todo};
+        #{ActionResId := #{mode := aggregated} = ActionState} ->
+            Batch = [Data || {_, Data} <- Batch0],
+            run_aggregated_action(Batch, ActionState)
+    end;
+on_batch_query(_ConnResId, Batch, _ConnState) ->
+    {error, {unrecoverable_error, {bad_batch, Batch}}}.
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+-spec insert_report(action_resource_id(), _Opts :: map()) -> {ok, map()} | {error, term()}.
+insert_report(ActionResId, Opts) ->
+    emqx_resource:simple_sync_query(
```

Comment:
why a loopback to self is needed?
because the http config is not in the context ?

Reply:
it's not self.  this is for debugging, not called by the connector itself.